libsolutil/DisjointSet.cpp

```diff
@@ -0,0 +1,109 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolutil/DisjointSet.h>
+
+#include <liblangutil/Exceptions.h>
+
+#include <numeric>
+
+using namespace solidity::util;
+
+ContiguousDisjointSet::ContiguousDisjointSet(size_t const _numNodes):
+	m_parents(_numNodes),
+	m_neighbors(_numNodes),
+	m_sizes(_numNodes, static_cast<value_type>(1)),
+	m_numSets(_numNodes)
+{
+	// each is their own neighbor and parent
+	std::iota(m_parents.begin(), m_parents.end(), 0);
+	std::iota(m_neighbors.begin(), m_neighbors.end(), 0);
+}
+
+size_t ContiguousDisjointSet::numSets() const { return m_numSets; }
+
+ContiguousDisjointSet::value_type ContiguousDisjointSet::find(value_type const _element) const
+{
+	solAssert(_element < m_parents.size());
+	// path halving
+	value_type rootElement = _element;
+	while (rootElement != m_parents[rootElement])
+	{
+		m_parents[rootElement] = m_parents[m_parents[rootElement]];
+		rootElement = m_parents[rootElement];
+	}
+	return rootElement;
+}
+
+void ContiguousDisjointSet::merge(value_type const _x, value_type const _y, bool const _mergeBySize)
```

Comment:
Are we planning to add rank in the future? Or is there a situation where we would have this false?
just a nitpick...

ps: I see it set to false in the test, but realistically, should be better to always have it on (or switched to rank), right?

Reply:
Good question:) The place we need it for now (loop detection with Tarjan's algorithm) relies on merging with "by size/rank" being switched off. That way a representative of a set can stay stable under merge operations. In particular this is used for loop headers (think a loop header as a vertex of a CFG represent a set of vertices that comprise the loop body). You can see this in [PR 15419](https://github.com/ethereum/solidity/pull/15419).

So generally I am inclined to agree that by rank is better - although potential problems with "by-size" are somewhat mitigated by path compression - but in our case it doesn't really matter. For now. :D

Reply:
Oh and there is just one test that has `mergeBySize` disabled, to check exactly that stability property of the set representative... the rest does have it enabled..

Reply:
Thanks for the explanation @clonker !